hasura.planx.uk/tests/lowcal_sessions.test.js

```diff
@@ -193,10 +254,42 @@ describe("lowcal_sessions", () => {
         expect(res.data.update_lowcal_sessions.returning).toHaveLength(1);
         expect(res.data.update_lowcal_sessions.returning[0].id).toEqual(bob1);
       });
+
+      test("Alice can update (but not select) her own existing session with an empty email ", async () => {
+        const headers = {
+          "x-hasura-lowcal-session-id": alice1,
+          "x-hasura-lowcal-email": ""
+        };
+        const res = await gqlPublic(updateByPK, { sessionId: alice1, data: { x: 1 } }, headers);
+        expect(res.data.update_lowcal_sessions_by_pk).toBeNull();
+      });
+
+      test("Anonymous users can upsert their own session with an empty email", async () => {
+        const headers = {
+          "x-hasura-lowcal-session-id": anon1,
+          "x-hasura-lowcal-email": ""
+        };
+
+        // inital insert (upsert)
+        const res1 = await gqlPublic(updateByPK, { sessionId: anon1, data: { x: 1 } }, headers);
+        expect(res1.data.update_lowcal_sessions_by_pk).not.toBeNull();
+        expect(res1.data.update_lowcal_sessions_by_pk.id).toEqual(anon1);
+        expect(res1.data.update_lowcal_sessions_by_pk.data).toHaveProperty("x", 1);
+
+        // update 1
+        const res2 = await gqlPublic(updateByPK, { sessionId: anon1, data: { y: 2 } }, headers);
+        expect(res2.data.update_lowcal_sessions_by_pk).not.toBeNull();
+        expect(res2.data.update_lowcal_sessions_by_pk.id).toEqual(anon1);
+        expect(res2.data.update_lowcal_sessions_by_pk.data).toHaveProperty("y", 2);
+
+        // update 2
+        const res3 = await gqlPublic(updateByPK, { sessionId: anon1, data: {} }, headers);
+        expect(res3.data.update_lowcal_sessions_by_pk).not.toBeNull();
+        expect(res3.data.update_lowcal_sessions_by_pk.data).toEqual({});
+      });
```

Comment:
I was surprised by the behavior here. I would have expected not to be able to read back updated data but I can only read what I have just written (and the deleted_at timestamp) and these tests show that is the case.

Reply:
So I think I worked this one out - we've modified `select_permissions` by adding the ` email: _neq: ""` check but not the `update_permissions` for this table.

This isn't a public user action, so I think it's fine to allow users to insert records without an email, but not allow them to update? This would mean that it's not possible for the public role to overwrite these audit records. Does that make sense?

Reply:
@DafyddLlyr if I remove permissions to update, we would only be able to insert a record capturing the payment information for the session once (because of the pk constraint). This would be an issue for cancelled and retried payments where the use also abandons the payment (the original scenario that triggered the work in this ticket) - there would be no way to know when the payment is going to be successful beforehand so no way to know which session to save before a potential abandonment scenario.

We've circled round this a few times and it sounds like you're uncomfortable with these public permissions for this particular edge case. My take is that a potential bad actor tampering with debug data in a Pay flow without a send component is not really a reasonable threat but I appreciate your zero-trust approach.

This could be an API feature if this approach isn't right. I'd also be happy to pass the baton to someone else in the team on this work if there's a better solution that I'm not seeing :runner: :runner:

Reply:
No thanks for clarifying that, you're right that the update is a requirement here.

I think we've both taken a detailed deep-dive here and understand the approach and with it's associated (very minor downsides).

I'm more than happy for this to proceed as is - the behaviour is well understood and documented with tests, and it's not something which can result in any practical risks. I just wanted to be sure we weren't missing anything here 👍 

✅ ✅ ✅